### PR TITLE
feat: NIP 04 Encrypted Direct Message

### PR DESCRIPTION
**Pull request by shanelau**

### CODE REVIEW

File: pubspec.yaml [hunk 1]
```diff
@@ -13,3 +13,5 @@ dependencies:
   bip340: ^0.1.0
   convert: ^3.1.1
   crypto: ^3.0.2
+  kepler: ^1.0.3
+  pointycastle:
```
Comment (ethicnology) [inline pubspec.yaml]: version missing ?

**Comment by ethicnology** [comment]
Hi @shanelau thank for your interest in contributing to this library.

You probably noticed someone else ask for this feature and send code to perform the NIP4 in https://github.com/ethicnology/dart-nostr/issues/15

1. The design of this library is to add fewer dependencies as possible, especially when it comes to cryptography, to reduce the attack surface. In your commit, you use BIP340/Kepler/crypto/pointycastle. BIP340 and Kepler are both wraping pointycastle. An interesting refactoring would be to kill these two dependencies to only use pointycastle. Even replacing crypto by pointycastle would be appreciated from a maintaining perspective.
2. Your current implementation should be reworked to be handy for users. Even without considering the 1st point, the actual usage of your work is not homogenous with the rest of the library. Could you take a look at how the [nip2](https://github.com/ethicnology/dart-nostr/blob/develop/lib/src/nips/nip_002.dart) has been implemented ? If you don't like this way, you can extend Event class with the encrypt/decrypt methods when kind=4
3. Where is the documentation ?
4. This library coverage is 98% currently and your commits does not include any unit test to validate your work

Please be sure any contribution is appreciated, but from my perspective, software engineering is not only about pushing code especially in open-source when we have to work together and understand each other, quality matter a lot.

If I have to take your code or the one given in https://github.com/ethicnology/dart-nostr/issues/15 and handling myself all the recommendations above, it would be easier for me to rewrite from scratch.

**Review by ethicnology — CHANGES_REQUESTED** [review]
Hi @shanelau thank for your interest in contributing to this library.

You probably noticed someone else ask for this feature and send code to perform the NIP4 in https://github.com/ethicnology/dart-nostr/issues/15

1. The design of this library is to add fewer dependencies as possible, especially when it comes to cryptography, to reduce the attack surface. In your commit, you use BIP340/Kepler/crypto/pointycastle. BIP340 and Kepler are both wraping pointycastle. An interesting refactoring would be to kill these two dependencies to only use pointycastle. Even replacing crypto by pointycastle would be appreciated from a maintaining perspective.
2. Your current implementation should be reworked to be handy for users. Even without considering the 1st point, the actual usage of your work is not homogenous with the rest of the library. Could you take a look at how the [nip2](https://github.com/ethicnology/dart-nostr/blob/develop/lib/src/nips/nip_002.dart) has been implemented ? If you don't like this way, you can extend Event class with the encrypt/decrypt methods when kind=4
3. Where is the documentation ?
4. This library coverage is 98% currently and your commits does not include any unit test to validate your work

Please be sure any contribution is appreciated, but from my perspective, software engineering is not only about pushing code especially in open-source when we have to work together and understand each other, quality matter a lot.

If I have to take your code or the one given in https://github.com/ethicnology/dart-nostr/issues/15 and handling myself all the recommendations above, it would be easier for me to rewrite from scratch.

**Comment by shanelau** [comment]
1. #15  is about fetch message kind=4
2. Q1: I removed the dependency reference of pointycastle. really don't know how to refactor
3. Q2： Has been adjusted to Class, encrypted message: `Nip4.getNip4Message(userPrivateKey, to, 'hello world');`
4. Q3: Github, Just Try 
5. Q4: Unit testing is indeed very important, sorry, there is really no time to complete it at this stage. Can be combined with #15  to test decrypted messages

**Comment by ntheden** [comment]
> 1. The design of this library is to add fewer dependencies as possible, especially when it comes to cryptography, to reduce the attack surface. In your commit, you use BIP340/Kepler/crypto/pointycastle. BIP340 and Kepler are both wraping pointycastle. An interesting refactoring would be to kill these two dependencies to only use pointycastle. Even replacing crypto by pointycastle would be appreciated from a maintaining perspective.

Regarding this point, the code in kepler and bip340 has to exist somewhere. Pointycastle has lower level code that those libraries build upon. I prototyped removing kepler, but it entailed basically copying its code into this project, which doesn't seem great to do. But in the case of kepler it might be worth it since that project is now archived.

Btw, removing `crypto` dependency was a one-liner:
```dart
-    List<int> hash = sha256.convert(utf8.encode(serializedEvent)).bytes;
+    Uint8List hash = SHA256Digest().process(Uint8List.fromList(utf8.encode(serializedEvent)));
```

**Pull request closed without merging**